### ISSUE TITLE
Entity.destroy(), Surface.create_entity()

### ISSUE THREAD

**Issue by aubergine10** [Afforess/Factorio-Stdlib#60]
Currently, there's no events fired if a mod destroys or creates an entity - and the devs have stated that they will not implement them. [Forum discussion](https://forums.factorio.com/viewtopic.php?f=34&t=34952)

As such, mods are going to need some consistent way to deal with this issue, and stdlib seems to be the ideal home for it.

Rough draft...

``` lua
-- both of these will need a generated event id
-- and a remote interface to get hold of those ids
local create_event
local destroy_event

function Entity.destroy_with_event( entity )
  game.raise_event( destroy_event, entity )
  entity.Destroy() -- this might fail, eg. rail with train on top cannot be destroyed
end

function Surface.create_entity_with_event( surface, tbl )
   local entity =  surface.create_entity( tbl )
   if entity then game.raise_event( create_event, entity ) end
end
```

**Comment by aubergine10** [comment]
Fleshed it out a bit more: https://github.com/aubergine10/lifecycle-events

**Comment by Afforess** [comment]
This is a good idea and solid start, but theoretically you can do even better than this. If entities have a trigger defined entity-spawn on death (used for biter spawners dropping alien artifacts), you can update prototypes to spawn an invisible entity on death, look for that spawn event, then use that as a general entity death notification. The only drawback is that prototype for the entity type that you want to notify upon death has to be updated before-hand. However, the advantage is that if the entity dies for any reason, in-game or scripted, it should trigger an event.

I think there should be an API to handle updating the prototype in the data-lifecycle phase and triggering the appropriate entity-death events from that as well.

**Comment by aubergine10** [comment]
By entity spawn on death do you mean "loot"? If so, that only happens when an entity `.die()` but not when it `.destroy()`. Also, the `.die()` method already triggers an event - the reason that `.die()` is unsuitable in most cases is that it causes all kinds of other things to happen - explosions, remnants, etc. The mods currently affected require immediate destruction of the entity (such as a vehicle put on to vehicle train, or creative mode clearing entities from an area)...

Also, by time loot is generated, entity has died - this is part of the problem the events are trying to solve. In mods that don't iterate lists of the entities they create/manage, knowing that "something" died is not sufficient - they need the actual entity so they can perform cleanup (eg. of related spawned entities or data).

If I've misunderstood what you were saying, maybe a pseudo-code example would help illustrate?

**Comment by Afforess** [comment]
I actually meant prototypes `attack_reaction`, which will trigger on damage or death. I suppose it has exactly the same issue as you've described as `loot` though.

**Comment by aubergine10** [comment]
I've never seen `attack_reaction` in any of the vanilla prototypes or mods - do you have some examples of it?

**Comment by Afforess** [comment]
There is some documentation of `attack_reaction` on [the factorio wiki](https://wiki.factorio.com/index.php?title=Types/AttackReaction), but I learned about it through [this thread](https://forums.factorio.com/viewtopic.php?f=28&t=25542).

**Comment by aubergine10** [comment]
Ah, nice, I can think of several uses for that already :)

IMO there still needs to be an optional event for .destroy() and .create_entity() to improve interoperability between mods.

**Comment by Nexela** [comment]
Closed with 536e9a7
